src/System.Text.Json/src/System/Text/Json/Node/JsonNode.Traversal.cs

```diff
@@ -0,0 +1,327 @@
+                        {
+                            recursionStack.Push(new KeyValuePair<string, JsonElement?>(property.Name, property.Value));
+                        }
+                        break;
+                    case JsonValueKind.Array:
+                        var jsonArray = new JsonArray();
+
+                        // Add jsonArray to current nodes:
+                        currentNodes.Push(new KeyValuePair<string, JsonNode>(currentPair.Key, jsonArray));
+
+                        // Add end of array marker:
+                        recursionStack.Push(new KeyValuePair<string, JsonElement?>(null, null));
+
+                        // Add elements to recursion stack. Reverse enumerator to keep items order:
+                        foreach (JsonElement element in currentJsonElement.Value.EnumerateArray().Reverse())
+                        {
+                            recursionStack.Push(new KeyValuePair<string, JsonElement?>(null, element));
+                        }
+                        break;
+                    case JsonValueKind.Number:
+                        var jsonNumber = new JsonNumber(currentJsonElement.Value.GetRawText());
+                        AddToParent(new KeyValuePair<string, JsonNode>(currentPair.Key, jsonNumber), ref currentNodes, ref toReturn);
+                        break;
+                    case JsonValueKind.String:
+                        var jsonString = new JsonString(currentJsonElement.Value.GetString());
+                        AddToParent(new KeyValuePair<string, JsonNode>(currentPair.Key, jsonString), ref currentNodes, ref toReturn);
+                        break;
+                    case JsonValueKind.True:
+                        var jsonBooleanTrue = new JsonBoolean(true);
+                        AddToParent(new KeyValuePair<string, JsonNode>(currentPair.Key, jsonBooleanTrue), ref currentNodes, ref toReturn);
+                        break;
+                    case JsonValueKind.False:
+                        var jsonBooleanFalse = new JsonBoolean(false);
+                        AddToParent(new KeyValuePair<string, JsonNode>(currentPair.Key, jsonBooleanFalse), ref currentNodes, ref toReturn);
+                        break;
+                    case JsonValueKind.Null:
+                        var jsonNull = new JsonNull();
+                        AddToParent(new KeyValuePair<string, JsonNode>(currentPair.Key, jsonNull), ref currentNodes, ref toReturn);
+                        break;
+                    default:
+                        Debug.Assert(jsonElement.ValueKind == JsonValueKind.Undefined, "No handler for JsonValueKind.{jsonElement.ValueKind}");
+                        throw ThrowHelper.GetJsonElementWrongTypeException(JsonValueKind.Undefined, jsonElement.ValueKind);
+                }
+            }
+
+            Debug.Assert(toReturn != null);
```

Comment:
Is this actually a failing scenario? What happens today if I do JsonNode.DeepCopy(JsonDocument.Parse(""))

Reply:
JsonDocument will throw JsonReaderException then.